crates/qasm2/src/lex.rs

```diff
@@ -0,0 +1,772 @@
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+//! The lexing logic for OpenQASM 2, responsible for turning a sequence of bytes into a
+//! lexed [TokenStream] for consumption by the parsing machinery.  The general strategy here is
+//! quite simple; for all the symbol-like tokens, the lexer can use a very simple single-byte
+//! lookahead to determine what token it needs to emit.  For keywords and identifiers, we just read
+//! the identifier in completely, then produce the right token once we see the end of the
+//! identifier characters.
+//!
+//! We effectively use a custom lexing mode to handle the version information after the `OPENQASM`
+//! keyword; the spec technically says that any real number is valid, but in reality that leads to
+//! weirdness like `200.0e-2` being a valid version specifier.  We do things with a custom
+//! context-dependent match after seeing an `OPENQASM` token, to avoid clashes with the general
+//! real-number tokenisation.
+
+use hashbrown::HashMap;
+use pyo3::prelude::PyResult;
+
+use std::path::Path;
+
+use crate::error::{message_generic, Position, QASM2ParseError};
+
+/// Tokenised version information data.  This is more structured than the real number suggested by
+/// the specification.
+#[derive(Clone, Debug)]
+pub struct Version {
+    pub major: usize,
+    pub minor: Option<usize>,
+}
+
+/// The context that is necessary to fully extract the information from a [Token].  This owns, for
+/// example, the text of each token (where a token does not have a static text representation),
+/// from which the other properties can later be derived.  This struct is effectively entirely
+/// opaque outside this module; the associated functions on [Token] take this context object,
+/// however, and extract the information from it.
+#[derive(Clone, Debug)]
+pub struct TokenContext {
+    text: Vec<String>,
+    lookup: HashMap<Vec<u8>, usize>,
+}
+
+impl TokenContext {
+    /// Create a new context for tokens.  Nothing is heap-allocated until required.
+    pub fn new() -> Self {
+        TokenContext {
+            text: vec![],
+            lookup: HashMap::new(),
+        }
+    }
+
+    /// Take ownership of the given `ascii_text` of a [Token], and return an index into the
+    /// [TokenContext].  This will not store strings that are already present in the context;
+    /// instead, the previous index is transparently returned.
+    fn index(&mut self, ascii_text: &[u8]) -> usize {
+        match self.lookup.get(ascii_text) {
+            Some(index) => *index,
+            None => {
+                let index = self.text.len();
+                self.lookup.insert(ascii_text.to_vec(), index);
+                self.text
+                    .push(std::str::from_utf8(ascii_text).unwrap().to_owned());
+                index
+            }
```

Comment:
I didn't originally do this because it didn't feel very nice to have the `with` closure have the massive side effect of doing `self.text.push(...)` - I tend to think of that type of closure as wanting to be a pure function.  If you think it's more idiomatic to use the closure with that side effect I can do it, though.

Reply:
I've voiced that same exact point in the past 😄. But these days, I'm more inclined to go for whatever has less code (assuming it's readable...). The match itself could be seen as a function that has a side-effect in this case. 

But, I actually don't feel strongly since the original is easy to understand as well.

Reply:
Ah, I went to go and do this for you, and remembered the other reason I didn't do it: `HashMap::entry` requires an owned key, which we don't have here unless we do a new allocation for every single look-up.  That's a massive hit on the runtime performance of the lexer.